### PR TITLE
Tokio v1.0.0 support.

### PR DESCRIPTION
**Pull request by mnbbrown**
Opening a PR incase it's helpful. New to rust so not sure if it's the best implementation.

Draft because:

- [ ] cleanup
- [ ] windows support?

Addresses #31 and #32

### CODE REVIEW

**Comment by wess** [comment]
Are you looking for someone to help out with your PR? I would love to get upstream using tokio 1.

**Comment by ivoisbelongtous** [comment]
I think with tokio's replacement of `PollEvented` with the Unix-specific `AsyncFd` there is no longer a clean way to integrate tokio and mio-serial on Windows. Perhaps Windows support should be dropped until a Windows equivalent of `AsyncFd` is implemented in tokio?

**Comment by jadamcrain** [comment]
I've talked with the Tokio folks on discord, and there is literally no way to hook an asynchronous serial port via  a windows `Handle` in 1.0.

The only solution to provide Windows support in this version of Tokio would be to run the windows port on a standard thread, and then tie it to the async resource using something like a DuplexStream:

https://docs.rs/tokio/1.0.1/tokio/io/struct.DuplexStream.html

Long term, we need to look into how Tokio implements the Windows iodriver to see if the comm port handle could be plugin via a similar wrapper like `AsyncFd`.

**Comment by jadamcrain** [comment]
BTW, I've raised an issue on Tokio so that we can coordinate with them to see if there is a path forward:

https://github.com/tokio-rs/tokio/issues/3396

File: src/lib.rs [hunk 1]
```diff
@@ -222,15 +222,13 @@ impl AsyncRead for Serial {
...
-                Ok(result) => return Poll::Ready(result),
-                Err(_would_block) => continue,
-            }
+        let mut guard = ready!(self.io.poll_read_ready(cx))?;
+        match guard.try_io(|_| {
+            let read = self.io.get_ref().read(buf.initialize_unfilled())?;
+            return Ok(buf.advance(read));
+        }) {
+            Ok(result) => return Poll::Ready(result),
+            Err(_would_block) => return Poll::Pending,
```
Comment (sjoerdsimons) [inline src/lib.rs:231]: This breaks as the `read()` can return would block in which case you go pending without the async fd actually registered with the runtime;  The original version with the loop is correct as `read()` returns would_block then the `poll_read_ready()` call should return Pending and everything should work nicely :).

fwiw i hit this problem in practise while playing with this branch and a fdti usb dongle in linux

**Comment by sjoerdsimons** [comment]
As the discussion on the tokio side for windows support is still ongoing; Is there a chance this could move foward without windows support as a step for e.g. tokio-serial 5.0 ?

**Comment by berkowski** [comment]
I'll finally have time to work on this, but first up is getting berkowski/mio-serial#23 resolved.

**Comment by Darksonn** [comment]
Hey. I am currently looking into the Tokio situation on windows. Would named pipes alone resolve the windows support, or do you need some sort of `AsyncHandle` beyond that to support it?

**Comment by jadamcrain** [comment]
Any updates on this @berkowski or @Darksonn?

Getting serial working again with Tokio on Windows would be something my firm would be willing to financially sponsor.

**Comment by Darksonn** [comment]
Named pipes are coming along nicely in https://github.com/tokio-rs/tokio/pull/3760, but I haven't heard anything from @berkowski in several months, so it's unclear what is going to be happening on the tokio-serial side.

File: src/lib.rs (hunk 1 above)
Comment (benjumanji) [reply src/lib.rs:231]: I'm confused about this: if the read returns EWOULDBLOCK doesn't that imply a bug somewhere else? We can't arrive at line 227 without being signaled by tokio that there are in fact _some_ bytes to read. I haven't been using this branch but I have some code that is morally equiv that I have been testing with ftdi dongles (snap!) and I haven't seen this condition triggered (yet).

Comment (Darksonn) [reply src/lib.rs:231]: It is possible to get a false positive from `poll_read_ready` and get the would block error, but the snippet is still incorrect — you have to loop around and call `poll_read_ready` again if that happens.

Comment (benjumanji) [reply src/lib.rs:231]: I see what you are saying. Thanks for making that connection. I also completely misread the sample code for `AsyncFd`. I guess I'd just gotten lucky :)

Comment (mateuszkj) [reply src/lib.rs:231]: @berkowski Please look at PR: https://github.com/mnbbrown/tokio-serial/pull/1

Comment (david-mcgillicuddy-moixa) [reply src/lib.rs:231]: I can confirm I have been seeing this bug too and rfael's patch fixed it.

**Comment by david-mcgillicuddy-moixa** [comment]
Named pipes just got released https://github.com/tokio-rs/tokio/releases/tag/tokio-1.7.0 🎊

**Comment by ColinFinck** [comment]
I've implemented the missing Windows part in https://github.com/enlyze/tokio-serial/tree/v4.4.0_serialstream_multiplatform
See also https://github.com/berkowski/tokio-serial/pull/39

This works well for me under Windows and Linux.
Requires pinning branches/revisions in Cargo.toml as long as they are not merged to master and released on crates.io. But other than that, I'm finally able to use serial ports with mio 0.7.x and tokio 1.x :)

**Comment by berkowski** [comment]
beta support for tokio 1.X available on crates.io with release 5.4.0-beta1.

**Pull request closed without merging**